benchmark/main.go

```diff
@@ -0,0 +1,342 @@
+package main
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/GaryBoone/GoStats/stats"
+)
+
+// Message describes a message
+type MessagePayload struct {
+	ID      string
+	Sent    time.Time
+	Payload interface{}
+}
+type Message struct {
+	ID             string
+	Topic          string
+	QoS            byte
+	Payload        MessagePayload
+	Sent           time.Time
+	Delivered      time.Time
+	DeliveredToSub time.Time
+	Error          bool
+}
+
+// RunResults describes results of a single client / run
+type RunResults struct {
+	ID        string `json:"id"`
+	Successes int64  `json:"successes"`
+	Failures  int64  `json:"failures"`
+
+	RunTime     float64 `json:"run_time"`
+	MsgTimeMin  float64 `json:"msg_time_min"`
+	MsgTimeMax  float64 `json:"msg_time_max"`
+	MsgTimeMean float64 `json:"msg_time_mean"`
+	MsgTimeStd  float64 `json:"msg_time_std"`
+
+	MsgDelTimeMin  float64 `json:"msg_del_time_min"`
+	MsgDelTimeMax  float64 `json:"msg_del_time_max"`
+	MsgDelTimeMean float64 `json:"msg_del_time_mean"`
+	MsgDelTimeStd  float64 `json:"msg_del_time_std"`
+
+	MsgsPerSec float64 `json:"msgs_per_sec"`
+}
+
+// TotalResults describes results of all clients / runs
+type TotalResults struct {
+	Ratio             float64 `json:"ratio"`
+	Successes         int64   `json:"successes"`
+	Failures          int64   `json:"failures"`
+	TotalRunTime      float64 `json:"total_run_time"`
+	AvgRunTime        float64 `json:"avg_run_time"`
+	MsgTimeMin        float64 `json:"msg_time_min"`
+	MsgTimeMax        float64 `json:"msg_time_max"`
+	MsgDelTimeMin     float64 `json:"msg_del_time_min"`
+	MsgDelTimeMax     float64 `json:"msg_del_time_max"`
+	MsgTimeMeanAvg    float64 `json:"msg_time_mean_avg"`
+	MsgTimeMeanStd    float64 `json:"msg_time_mean_std"`
+	MsgDelTimeMeanAvg float64 `json:"msg_del_time_mean_avg"`
+	MsgDelTimeMeanStd float64 `json:"msg_del_time_mean_std"`
+	TotalMsgsPerSec   float64 `json:"total_msgs_per_sec"`
+	AvgMsgsPerSec     float64 `json:"avg_msgs_per_sec"`
+}
+
+// JSONResults are used to export results as a JSON document
+type JSONResults struct {
+	Runs   []*RunResults `json:"runs"`
+	Totals *TotalResults `json:"totals"`
+}
+
+// Connection represents connection
+type Connection struct {
+	ChannelID string `json:"ChannelID"`
+	ThingID   string `json:"ThingID"`
+	ThingKey  string `json:"ThingKey"`
+	MtlsCert  string `json:"MtlsCert"`
+	MtlsKey   string `json:"MtlsKey"`
+}
+
+func main() {
+
+	var (
```

Comment:
We use [Cobra](https://github.com/spf13/cobra) to create our CLI tools (take a look at SDK for example). It helps you read input config params and well organizing your code for functions that implement the logic.

We use [Viper](https://github.com/spf13/viper) to read various config files - in this case it will be in TOML.

Reply:
@drasko I agree with you, but I can't help but feel that this is cleaner and simpler way of doing the same thing (at least when it comes to the bare evaluation of env vars; these libs are doing much more, ofc).

Reply:
cobra and viper added